pkg/cmd/rollbackcopy/etcdclientutils.go

```diff
@@ -0,0 +1,137 @@
+package rollbackcopy
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"google.golang.org/grpc"
+	"io"
+	"k8s.io/klog"
+	"os"
+	"time"
+
+	"go.etcd.io/etcd/clientv3"
+	"go.etcd.io/etcd/pkg/transport"
+)
+
+func getEtcdClient(endpoints []string) (*clientv3.Client, error) {
+	dialOptions := []grpc.DialOption{
+		grpc.WithBlock(), // block until the underlying connection is up
+	}
+
+	tlsInfo := transport.TLSInfo{
+		CertFile:      os.Getenv("ETCDCTL_CERT"),
+		KeyFile:       os.Getenv("ETCDCTL_KEY"),
+		TrustedCAFile: os.Getenv("ETCDCTL_CACERT"),
+	}
+	tlsConfig, err := tlsInfo.ClientConfig()
+
+	cfg := &clientv3.Config{
+		DialOptions: dialOptions,
+		Endpoints:   endpoints,
+		DialTimeout: 2 * time.Second,
+		TLS:         tlsConfig,
+	}
+
+	cli, err := clientv3.New(*cfg)
+	if err != nil {
+		return nil, fmt.Errorf("failed to make etcd client for endpoints %v: %w", endpoints, err)
+	}
+	return cli, nil
+}
+
+func getClusterVersionAndUpgradeInfo(cli *clientv3.Client) (string, bool, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	clusterversionKey := "/kubernetes.io/config.openshift.io/clusterversions/version"
+	resp, err := cli.Get(ctx, clusterversionKey)
+	if err != nil {
+		return "", false, err
+	}
+	klog.Info("### Response for client GET: ", resp)
+	if len(resp.Kvs) != 1 {
+		klog.Errorf("Expected to get a single key from etcd, got %d", len(resp.Kvs))
+		return "", false, fmt.Errorf("Expected to get a single key from etcd, got %d", len(resp.Kvs))
+	}
+
+	klog.Info("### Value lookup: ", string(resp.Kvs[0].Value))
+	var cv map[string]interface{}
+	if err := json.Unmarshal(resp.Kvs[0].Value, &cv); err != nil {
+		return "", false, err
+	}
+
+	status := cv["status"].(map[string]interface{})
+	desired := status["desired"].(map[string]interface{})
+	klog.Info("desired version: ", desired["version"])
+
+	history := status["history"].([]interface{})
+	latestHistory := history[0].(map[string]interface{})
+	klog.Info("latest history version:", latestHistory["version"], " status: ", latestHistory["state"])
+	klog.Info("Return values: ", latestHistory["version"], " upgradeInProgress: ", desired["version"] != latestHistory["version"])
```

Comment:
I wouldn't log anything at all in this function unless necessary, or increase V level if you want to keep